### ISSUE TITLE
There is some racing happening

### ISSUE THREAD

**Issue by mokhosh** [joshhanley/sprucewire#2]
https://user-images.githubusercontent.com/6499685/108670300-e69f3300-74f3-11eb-810f-3911384acc5f.mov

As you can see in this screen recording, when I type or when I clear what I typed, it rewrites the value over and over. Sometimes it looks a bit worse than this.

**Comment by stancl** [comment]
I think Sprucewire should keep received messages in a blocking FIFO queue that requires them to be executed in a specific order.

An alternative would be to just drop responses older than the already received ones, but maybe that could miss out on some state or browser events?

**Comment by mokhosh** [comment]
I don't know if this is the right direction, but I always thought Livewire should make a difference between an update that was initiated bu user input, and updates that are initiated by Livewire.

That distinction could potentially avoid and solve a lot of issues.

**Comment by stancl** [comment]
> Livewire should make a difference between an update that was initiated bu user input, and updates that are initiated by Livewire

When you click a button with `wire:click="foo"`, it's as if you clicked a button with `@click="$wire.foo()`, which is as if you did `Livewire.find('something').call('foo')`.

**Comment by joshhanley** [comment]
Similar issue was reported on the demo repo. I've copied my answer below.

I'd love to work out a solution for it, as I think it could improve the experience for Livewire too.

It's like you still need to handle the Livewire returned request, but if the property is now dirty, don't overwrite it from the returned request.

That issue there is Livewire on backend could potentially modify the property to something completely different, in which you don't want to loose that change.

Would love any ideas around that!

---

Yeah that's tough as it's how Livewire handles returned requests. I will have to see if there is a way I can detected that.

I'd recommend using wire:model.defer or wire:model.lazy as much as possible, which prevents that issue. Or even adding a debounce.

You get similar issues with Livewire with input flickering when not using defer or lazy.
livewire/livewire#724
livewire/livewire#1007

Only difference here is when it "flickers" its triggering updates for child component, and then back to the parent component, causing the infinite loop.

**Comment by mokhosh** [comment]
> When you click a button with wire:click="foo", it's as if you clicked a button with @click="$wire.foo(), which is as if you did Livewire.find('something').call('foo').

@stancl Yes, that is the current behaviour. But can we create a layer in between where every change in state happens, and where we can keep track of more information about the changes?

@joshhanley Thanks for giving the background of the issue. I still think the core of the issue is finding a way to differentiate between who is changing state and why.
